### DISCUSSION TITLE
Vimix Chataigne Module

### DISCUSSION THREAD

**Discussion by lupin3rd** [brunoherbelin/vimix#119] [Show and tell]
I release a Chataigne module to use Vimix.
I hope it's useful to someone.
https://github.com/lupin3rd/vimix-chataigne-module

**Comment by brunoherbelin** [comment]
GREAT ! :smile_cat:  
Lots of improvements thanks to you Lupin3rd since #7 ! This was over 3 years ago, already!!


This will definitely be useful for mapping MIDI to vimix ! See #116 
You could share a simple demo Chataigne file? I could add a section in the Documentation.

**Comment by lupin3rd** [comment]
YEAH! :+1: 

Sure.  I could share a demo and some screenshots.

**Comment by lupin3rd** [comment]
Sorry for my bad english... a little test:
https://www.youtube.com/watch?v=gfelSTZ6fpg

**Reply by brunoherbelin** [reply]
This is great !! I will add a section in the Wiki to document the interfacing with Chataigne !
Can I use this video as example ?

**Reply by lupin3rd** [reply]
Sure, despite my bad english :-)
I would try to do some good videos with midi mapping, interaction with music, etc.

**Reply by brunoherbelin** [reply]
Already very nice for a simple wiki page : https://github.com/brunoherbelin/vimix/wiki/OSC-Mapping

**Comment by lupin3rd** [comment]
@brunoherbelin , we some OSC message:
- enable/disable Color Correction;
- send Gamma values;
- inactive a source, actually we have stop (play < 0.5) and pause that do same things; and don't have a message to stop (incative a source), also alpha = 0 don't inactive a source.

**Reply by brunoherbelin** [reply]
It will be my pleasure to add the OSC commands needed.
Can you list below what you suggest? (including the ones commented elsewhere) I can work on this next week.

**Comment by brunoherbelin** [comment]
Updated OSC documentation at https://github.com/brunoherbelin/vimix/wiki/Open-Sound-Control-API
And associated updates in Beta e42afcb4343735d7eb90e68856792dcf3ebf0a83

**Reply by lupin3rd** [reply]
I've updated Vimix-Chataigne-module with OSC messages of your last commit.
Some bugs that i find:
- Gamma OSC message 0 to 1 change gamma value on vimix from -1 to 0
- If i create a text source empty in Vimix, vimix crash.

Some advices about OSC messages to add:
- inactive a source, actually we have stop (play < 0.5) and pause that do same things; and don't have a message to stop (incative a source), also alpha = 0 don't inactive a source;
- OSC messages to interact with filter of a cloned source;
- OSC messages to interact with gstreamer effects.

Some advices not correlated to OSC:
- i think that in vimix GUI the color corrections values section must be always visible (also with color correction off); in this way we can change  values and then enable color correction;
- i think that sources icons on left bar are useless (only 2 chars don't suggest a source); it would be nice an image thumb;
- i think that can be usefull a mouseover text with name of source when i go over a source, without change current;
- change the position of "gstreamer effect" form videoplayer timeline to source panel (with color correction and filters) to uniform the GUI.

**Reply by brunoherbelin** [reply]
>    Gamma OSC message 0 to 1 change gamma value on vimix from -1 to 0
>    If i create a text source empty in Vimix, vimix crash.

Thanks, both bugs are fixed in Beta 01e21ea2124a28fd346ab928671a874a59a67d28

> inactive a source, actually we have stop (play < 0.5) and pause that do same things

yes indeed, it is on purpose to allow easy mapping with OSC in opposite ways; some clients want to send button press to play, and others send button press to pause. This is needed for some OSC clients that are not allowing to script.

> don't have a message to stop (incative a source)

Well, not sure to get what you need: an inactive source is not playing, so sending play to TRUE will not do anything...

> alpha = 0 don't inactive a source

Yes and it is the intended behavior: alpha corresponds to the placement in the mixing ring : a source at the border of the purple ring (alpha = 0) is still active: to make it inactive, it should be outside of the darker circle, i.e with a negative alpha. Send an OSC alpha to a negative value to deactivate it (e.g. `/vimix/#0/alpha f -0.5`)

> OSC messages to interact with filter of a cloned source

yep, this could be done, but the name of the parameter depends on the type of filter (e.g. 'radius' for blur, 'amount' for sharpen). so its not so clear. Maybe just `/vimix/#0/filtervalue` ?

> OSC messages to interact with gstreamer effects

okay, i didn't want to emphasize this feature but it could be added

>  i think that in vimix GUI the color corrections values section must be always visible (also with color correction off); in this way we can change values and then enable color correction

Humm.. I see your point. Also in the GUI it should be different then, with an enable/disable switch which does not close the list of values. I will investigate.

> i think that sources icons on left bar are useless (only 2 chars don't suggest a source); it would be nice an image thumb

Well, that's a design choice : the initials are here to refer to the name of the source, that is intended to be the way to refer to sources. It is the name / initials that are used to refer to the source in menus (e.g. action mapping). Still, I can understand that one could work more on the visual aspect, and in practice it is possible to program an icon that shows the current frame.

The problems with showing the source image frame:
- sometimes the source is in a state that does not represent what it is : an empty frame, black or transparent, e.g. paused at a black frame. Showing this as the source icon does not help (you would have a list of black rectangles on the left bar)
- the size of the thumbnail does not allow seeing what it is, being too small
- the thumbnail aspect ratio is not square (but i could cut to square as thumbnail preview)

Still, taking into account your request to have more informative icons for sources in the left panel, we could think of
- mouse-over the letters shows a tooltip with preview of the source current frame
- small icon in the lower right corner of the letters icon to  indicate the type of the source (same icon of the source in mixing view)
- alternatively, a user-defined frame to use as icon : e.g. a camera icon at the right of the preview in the panel with 'clic to take screenshot as thumbnail' ... to be checked

> i think that can be usefull a mouseover text with name of source when i go over a source, without change current;

yes, same as proposed above about preview on mouse over


> change the position of "gstreamer effect" form videoplayer timeline to source panel (with color correction and filters) to uniform the GUI.

This is simply not possible : the gstreamer effect is specific to media source and cannot be added to others.

**Reply by lupin3rd** [reply]
> Yes and it is the intended behavior: alpha corresponds to the placement in the mixing ring : a source at the border of the purple ring (alpha = 0) is still active: to make it inactive, it should be outside of the darker circle, i.e with a negative alpha. Send an OSC alpha to a negative value to deactivate it (e.g. /vimix/#0/alpha f -0.5)

Good. I never try to send a negative alpha because in API documentation i read that alpha is from 0.0 to 1.0.

> yep, this could be done, but the name of the parameter depends on the type of filter (e.g. 'radius' for blur, 'amount' for sharpen). so its not so clear. Maybe just /vimix/#0/filtervalue ?

We need to find a solution that work with filter that have a parameter and also with filter that have more parameters (like blur).
Maybe /vimix/#0/filtername (0 to 1) to enable a filter and then /vimix/#0/filtername/param1 (0 to 1) , /vimix/#0/filtername/param2 (0 to 1) set param2 if exist, etc

**Reply by brunoherbelin** [reply]
> Good. I never try to send a negative alpha because in API documentation i read that alpha is from 0.0 to 1.0.

Indeed documentation was outdated! Updated now.

**Reply by lupin3rd** [reply]
About filter question: in the GUI actually is not possible to save filter that can be created by editor; it's important add filter by user, or save and reopen.

**Comment by brunoherbelin** [comment]
Check out Beta 2c7262ced4dfbe79d8cb1db6915b5ee73ed386fb : playing with a tooltip on source left icons:

![image](https://github.com/brunoherbelin/vimix/assets/60393189/5cbaa7e8-d631-4088-aa3d-ef2e13656e55)

**Reply by lupin3rd** [reply]
This is a very good solution ! I tested. Works very fine.

**Reply by lupin3rd** [reply]
I suggest to add also a mouse-over the source on table that show a tooltip with the name and ID of the source.

**Comment by MoorPoor** [comment]
Hei!
Just tried to find the filter section in chataigne module... 
Thank you for already investigating in this direction!
great combination of software!

**Comment by lupin3rd** [comment]
Hi @MoorPoor  and @brunoherbelin
I report below the ideas regarding the filter section that came up in the discussion, so we can alayze and discute:
- We need to find a solution that work with filter that have a parameter and also with filter that have more parameters (like blur).
Maybe /vimix/#0/filtername (0 to 1) to enable a filter and then /vimix/#0/filtername/param1 (0 to 1) , /vimix/#0/filtername/param2 (0 to 1) set param2 if exist, etc
- in the GUI actually is not possible to save filter that can be created by editor; it's important add filter by user, or save and reopen. And use those filters by OSC like others (we can think about a filter directory with some filtername.txt that contain filter's code, and vimix load all those filter when start)
- we need to interact with filter's vars by OSC. Like do GlslViewer (i use it and capture output with vimix, in this way we can do interaction):
https://github.com/patriciogonzalezvivo/glslViewer/wiki/Interacting-with-GlslViewer#osc-commands

what do you think?

**Comment by brunoherbelin** [comment]
Excited that we can work on this! Important to have a discussion to design things right :smile: 

> in the GUI actually is not possible to save filter that can be created by editor; it's important add filter by user, or save and reopen. And use those filters by OSC like others (we can think about a filter directory with some filtername.txt that contain filter's code, and vimix load all those filter when start)

I am not sure to understand the 'not possible to save filter' : code is saved in the session file, with the clone source.
Still, as I agree it could be useful to load / save code, I added the import and export in the GUI (still in Beta):

![image](https://github.com/brunoherbelin/vimix/assets/60393189/aa8ef014-1d76-4e3c-94b6-47e28579059c)

> We need to find a solution that work with filter that have a parameter and also with filter that have more parameters (like blur).
> Maybe /vimix/#0/filtername (0 to 1) to enable a filter and then /vimix/#0/filtername/param1 (0 to 1) , /vimix/#0/filtername/param2 (0 to 1) set param2 if exist, etc

Indeed, and I could propose the following OSC API:

    /vimix/#0/filter sff "blur" 0.3 1.2

 - `s` for string, with filter name 
 - `f` for float, as many as needed 

The above would mean set filter to "blur", with those two values as options.
Also: remember OSC allows for the `N` null type, which thus allows to do this:

    /vimix/#0/filter NNf 1.54

This above would mean to keep same `s` (not change filter name), keep same first float, and change only second float.

> we need to interact with filter's vars by OSC. Like do GlslViewer (i use it and capture output with vimix, in this way we can do interaction):

I have a problem with this, as declaring custom uniforms is not Shadertoy compatible. Instead, a hack is to use the `iMouse` generic uniform to modify values with the GUI:

![image](https://github.com/brunoherbelin/vimix/assets/60393189/276e9016-8af5-47a1-89a8-459cc510f990)

However, I agree this is not optimal. I am not sure of how to proceed. I will check how GLSLViewer works.

**Reply by lupin3rd** [reply]
> I am not sure to understand the 'not possible to save filter' : code is saved in the session file, with the clone source.

I never saved a cloned source :-) my lacks. Very good.

> Still, as I agree it could be useful to load / save code, I added the import and export in the GUI (still in Beta)

Good. I will test.

Your propose about filter's OSC messages is good for me!

**Comment by brunoherbelin** [comment]
Hey! Please try 8e28eba9595b304658c6ab122096374eedcac9a7 Beta : I added the Filter parameter in OSC.
Examples:

To just set filter by name:

    /vimix/current/filter s blur


To set filter and a specific method:

    /vimix/current/filter ss blur fast


To set filter, method, and value:

    /vimix/current/filter ssf blur gaussian 1.0


To set filter, method, and animate value change in 500ms:

    /vimix/current/filter ssff blur gaussian 1.0 500

And indicate `N` in the list of parameter types to ignore the filter name or method:

    /vimix/current/filter NNff 1.0 500
    /vimix/current/filter Ns gaussian



Would be nice to have your feedback

**Comment by brunoherbelin** [comment]
NB: I updated the [OSC API documentation](https://github.com/brunoherbelin/vimix/wiki/Open-Sound-Control-API) with more details.

**Reply by lupin3rd** [reply]
I'm testing and updating my Chataigne module... seem that there are two problems:
- float of delay don't seem to works;
- second float value (iterations) in scattered (blur) don't seem to works;

**Reply by brunoherbelin** [reply]
Thanks for testing and updating the Chataigne module :)

> float of delay don't seem to works

Here, `/vimix/#1/filter sNf delay 0.2` works fine...

> second float value (iterations) in scattered (blur) don't seem to works;

..and it is not meant to work like that : only one value is possible; the reason is that there is no way to distinguish what float means what with list of osc arguments. 

If you think it is absolutely necessary what can be done is that there is a third `s` argument to specify the name of the parameter, e.g. 

      /vimix/#1/filter sssff blur gaussian radius 0.2 400

I thought it would be overkill, and chose to have only one parameter exposed to OSC...

**Reply by lupin3rd** [reply]
Now i understand. I fixes all problems in module.
My last test:
- blur/fast OSC message don't work;
- in API documentation "edge" method of Edge must be "thresholding".
I commit new version soon.

**Reply by lupin3rd** [reply]
https://github.com/lupin3rd/vimix-chataigne-module/releases/tag/v0.1

**Reply by MoorPoor** [reply]
thank you for your efforts!
Unfortunately, I'm not able to compile any of the beta versions at the moment, as various errors keep popping up.
I'll try again in a few days.

**Comment by brunoherbelin** [comment]
Here is something which should be answering the needs discussed here : vimix is now able to recognise the declaration of a uniform variable in a filter and add it to the list of parameters in the GUI and in OSC.

Here is a demo in the GUI ;

https://github.com/brunoherbelin/vimix/assets/60393189/b5b262e1-5d7f-4e70-bec6-1a55dbdea5cf

For OSC, you simply set the `uniform` variable to a value with

       /vimix/current/uniform sf Size 0.4


Check the [documentation of OSC API](https://github.com/brunoherbelin/vimix/wiki/Open-Sound-Control-API#attributes-to-control-different-types-of-sources) for more details.

**Reply by brunoherbelin** [reply]
Should be working with Beta 133fb661b3388b49ed1bb17bb5f4692629f9ee01

**Reply by lupin3rd** [reply]
This is wonderfull! I tomorrow will test this...and update the chataigne module...

**Reply by lupin3rd** [reply]
I finally works on chataigne modulo and test :-)
All OSC messages now works fine.
Remain only one problem: if i use a wrong variable name i see a new slider in the GUI!
See this video.

https://github.com/brunoherbelin/vimix/assets/1455316/5038f610-cc90-41aa-a0fc-dbe70a1e02bc

**Reply by brunoherbelin** [reply]
Excellent ! that's great that the connection with Chataigne is now very complete and well integrated!

Regarding your comment: I don't think it's a bug, as by definition, a wrong variable name cannot be understood without reason. However, I admit vimix GUI is not very informative and there is no feedback to the user to help find the solution. I will work on an error message or something.

**Comment by lupin3rd** [comment]
I do some updates. Now the chataigne module receive vimix OSC messages:
https://github.com/lupin3rd/vimix-chataigne-module
